File: sprint3/By languages/JS/lesson12code1.js
function partition(array, pivot) {
  let left = [], center = [], right = [];
  for (let x of array) {
    if (x < pivot) {
      left.push(x);
    } else if (x === pivot) {
      center.push(x);
    } else {
      right.push(x);
    }
  }
  return [left, center, right];
}

function quicksort(array) {
  if (array.length < 2) {
    return array; // массивы с 0 или 1 элементами фактически отсортированы
  } else {
    let pivot = array[Math.floor(Math.random() * array.length)];
    let [left, center, right] = partition(array, pivot);
    return quicksort(left).concat(center).concat(quicksort(right));
  }
}

File: sprint3/By languages/python/lesson12code1.py
def partition(array, pivot):
    left = [x for x in array if x < pivot]
    center = [x for x in array if x == pivot]
    right = [x for x in array if x > pivot]
    return left, center, right

def quicksort(array):
    if len(array) < 2:
        return array   # массивы с 0 или 1 элементами фактически отсортированы
    else:
        pivot = random.choice(array)
        left, center, right = partition(array, pivot)
        return quicksort(left) + center + quicksort(right)
